In [7]:
import pandas as pd
import numpy as np
from IPython.display import display

# **Load Dataset**

In [8]:
customers           = pd.read_csv(r"olist_customers_dataset.csv")
geolocation         = pd.read_csv(r"olist_geolocation_dataset.csv")
items               = pd.read_csv(r"olist_order_items_dataset.csv")
payment             = pd.read_csv(r"olist_order_payments_dataset.csv")
o_reviews           = pd.read_csv(r"olist_order_reviews_dataset.csv")
orders              = pd.read_csv(r"olist_orders_dataset.csv")
products            = pd.read_csv(r"olist_products_dataset.csv")
product_category    = pd.read_csv(r"product_category_name_translation.csv")
sellers             = pd.read_csv(r"olist_sellers_dataset.csv")

# **Automated Exploratory Data Analysis (EDA)**
## **Loop through all datasets to check for quality issues (nulls, types, duplicates)**

In [ ]:

dataset = {
    "customers": customers,
    "geolocation": geolocation,
    "orders_items" : items,
    "payment": payment,
    "order_reviews": o_reviews,
    "orders": orders,
    "products": products,
    "product_category": product_category,
    "sellers": sellers
}

for name, df in dataset.items():
    print("\n" + "=" *150)
    print(f'Dataset: {name}')
    print("=" *150)

    # Display dataset dimensions
    print("\n [1] Shape")
    shape = pd.DataFrame(
        {"N_Rows": [df.shape[0]], "N_columns":[df.shape[1]]}
    )
    display(shape)

    # Check for identical rows
    print("\n [2] Check Duplicated")
    dupl = df.duplicated().sum()
    display(dupl)

    # Check column data types 
    print("\n [3] Data Type")
    columns_type = pd.DataFrame(
        {"Columns": df.columns, "Data Type": df.dtypes}
    )
    display(columns_type)

    # Statistical overview for numeric and categorical data
    print("\n [4] Describe")
    describe_df = df.describe(include= "all").T
    display(describe_df)
    
    # Identify missing values per column
    print("\n [5] Missing Data")
    null = pd.DataFrame({
        "Columns": df.columns, 
        "Missing Data": df.isnull().sum().values
    })
    display(null)

    # First 10 Rows
    print("\n [6] First Rows")
    display(df.head(10))

    # Last 10 Rows
    print("\n [7] Last Rows")
    display(df.tail(10))
    
    # Summary
    print("\n [8] Summary")
    summary = pd.DataFrame({
        "N_Rows": [df.shape[0]],
        "N_Columns": [len(df.columns)],
        "Memory": [df.memory_usage(deep= True).sum()]
    }) 
    display(summary)


Dataset: customers

 [1] Shape


N_Rows  N_columns
0   99441          5


 [2] Check Duplicated


np.int64(0)


 [3] Data Type


Columns Data Type
customer_id                            customer_id    object
customer_unique_id              customer_unique_id    object
customer_zip_code_prefix  customer_zip_code_prefix     int64
customer_city                        customer_city    object
customer_state                      customer_state    object


 [4] Describe


count unique                               top  \
customer_id                 99441  99441  06b8999e2fba1a1fbc88172c00ba8bc7   
customer_unique_id          99441  96096  8d50f5eadf50201ccdcedfb9e2ac8455   
customer_zip_code_prefix  99441.0    NaN                               NaN   
customer_city               99441   4119                         sao paulo   
customer_state              99441     27                                SP   

                           freq          mean           std     min      25%  \
customer_id                   1           NaN           NaN     NaN      NaN   
customer_unique_id           17           NaN           NaN     NaN      NaN   
customer_zip_code_prefix    NaN  35137.474583  29797.938996  1003.0  11347.0   
customer_city             15540           NaN           NaN     NaN      NaN   
customer_state            41746           NaN           NaN     NaN      NaN   

                              50%      75%      max  
customer_id                   NaN      NaN      NaN  
customer_unique_id            NaN      NaN      NaN  
customer_zip_code_prefix  24416.0  58900.0  99990.0  
customer_city                 NaN      NaN      NaN  
customer_state                NaN      NaN      NaN


 [5] Missing Data


Columns  Missing Data
0               customer_id             0
1        customer_unique_id             0
2  customer_zip_code_prefix             0
3             customer_city             0
4            customer_state             0


 [6] First Rows


customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
5  879864dab9bc3047522c92c82e1212b8  4c93744516667ad3b8f1fb645a3116a4   
6  fd826e7cf63160e536e0908c76c3f441  addec96d2e059c80c30fe6871d30d177   
7  5e274e7a0c3809e14aba7ad5aae0d407  57b2a98a409812fe9618067b6b8ebe4f   
8  5adf08e34b2e993982a47070956c5c65  1175e95fb47ddff9de6b2b06188f7e0d   
9  4b7139f34592b3a31687243a302fa75b  9afe194fb833f79e300e37e580171f22   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP  
5                     89254         jaragua do sul             SC  
6                      4534              sao paulo             SP  
7                     35182                timoteo             MG  
8                     81560               curitiba             PR  
9                     30575         belo horizonte             MG


 [7] Last Rows


customer_id                customer_unique_id  \
99431  be842c57a8c5a62e9585dd72f22b6338  ca186065de6e2d01cfc99763e6d62048   
99432  f255d679c7c86c24ef4861320d5b7675  d111b06b6f3a2add0d2241325f65b5ca   
99433  14308d2303a3e2bdf4939b86c46d2679  e7f8760e2bbd2f1986bebd99596c088e   
99434  f5a0b560f9e9427792a88bec97710212  b3e53d18a997f27a3ffd16da497eaf58   
99435  7fe2e80252a9ea476f950ae8f85b0f8f  4b5820135d360a45552b5163835b1d89   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix    customer_city customer_state  
99431                     99150            marau             RS  
99432                     13500        rio claro             SP  
99433                     66033            belem             PA  
99434                      7790          cajamar             SP  
99435                     35500      divinopolis             MG  
99436                      3937        sao paulo             SP  
99437                      6764  taboao da serra             SP  
99438                     60115        fortaleza             CE  
99439                     92120           canoas             RS  
99440                      6703            cotia             SP


 [8] Summary


N_Rows  N_Columns    Memory
0   99441          5  27877866


Dataset: geolocation

 [1] Shape


N_Rows  N_columns
0  1000163          5


 [2] Check Duplicated


np.int64(261831)


 [3] Data Type


Columns Data Type
geolocation_zip_code_prefix  geolocation_zip_code_prefix     int64
geolocation_lat                          geolocation_lat   float64
geolocation_lng                          geolocation_lng   float64
geolocation_city                        geolocation_city    object
geolocation_state                      geolocation_state    object


 [4] Describe


count unique        top    freq  \
geolocation_zip_code_prefix  1000163.0    NaN        NaN     NaN   
geolocation_lat              1000163.0    NaN        NaN     NaN   
geolocation_lng              1000163.0    NaN        NaN     NaN   
geolocation_city               1000163   8011  sao paulo  135800   
geolocation_state              1000163     27         SP  404268   

                                     mean          std         min        25%  \
geolocation_zip_code_prefix  36574.166466  30549.33571      1001.0    11075.0   
geolocation_lat                -21.176153     5.715866  -36.605374 -23.603546   
geolocation_lng                -46.390541     4.269748 -101.466766 -48.573172   
geolocation_city                      NaN          NaN         NaN        NaN   
geolocation_state                     NaN          NaN         NaN        NaN   

                                   50%        75%         max  
geolocation_zip_code_prefix    26530.0    63504.0     99990.0  
geolocation_lat             -22.919377  -19.97962   45.065933  
geolocation_lng             -46.637879 -43.767709  121.105394  
geolocation_city                   NaN        NaN         NaN  
geolocation_state                  NaN        NaN         NaN


 [5] Missing Data


Columns  Missing Data
0  geolocation_zip_code_prefix             0
1              geolocation_lat             0
2              geolocation_lng             0
3             geolocation_city             0
4            geolocation_state             0


 [6] First Rows


geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   
5                         1012       -23.547762       -46.635361   
6                         1047       -23.546273       -46.641225   
7                         1013       -23.546923       -46.634264   
8                         1029       -23.543769       -46.634278   
9                         1011       -23.547640       -46.636032   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP  
5        são paulo                SP  
6        sao paulo                SP  
7        sao paulo                SP  
8        sao paulo                SP  
9        sao paulo                SP


 [7] Last Rows


geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
1000153                        99970       -28.343273       -51.873734   
1000154                        99950       -28.070493       -52.011342   
1000155                        99965       -28.180655       -52.034367   
1000156                        99950       -28.072188       -52.011272   
1000157                        99950       -28.068864       -52.012964   
1000158                        99950       -28.068639       -52.010705   
1000159                        99900       -27.877125       -52.224882   
1000160                        99950       -28.071855       -52.014716   
1000161                        99980       -28.388932       -51.846871   
1000162                        99950       -28.070104       -52.018658   

        geolocation_city geolocation_state  
1000153          ciriaco                RS  
1000154         tapejara                RS  
1000155       agua santa                RS  
1000156         tapejara                RS  
1000157         tapejara                RS  
1000158         tapejara                RS  
1000159   getulio vargas                RS  
1000160         tapejara                RS  
1000161  david canabarro                RS  
1000162         tapejara                RS


 [8] Summary


N_Rows  N_Columns     Memory
0  1000163          5  136592627


Dataset: orders_items

 [1] Shape


N_Rows  N_columns
0  112650          7


 [2] Check Duplicated


np.int64(0)


 [3] Data Type


Columns Data Type
order_id                        order_id    object
order_item_id              order_item_id     int64
product_id                    product_id    object
seller_id                      seller_id    object
shipping_limit_date  shipping_limit_date    object
price                              price   float64
freight_value              freight_value   float64


 [4] Describe


count unique                               top  freq  \
order_id               112650  98666  8272b63d03f5f79c56e9e4120aec44ef    21   
order_item_id        112650.0    NaN                               NaN   NaN   
product_id             112650  32951  aca2eb7d00ea1a7b8ebd4e68314663af   527   
seller_id              112650   3095  6560211a19b47992c3666cc44a7e94c0  2033   
shipping_limit_date    112650  93318               2017-07-21 18:25:23    21   
price                112650.0    NaN                               NaN   NaN   
freight_value        112650.0    NaN                               NaN   NaN   

                           mean         std   min    25%    50%    75%     max  
order_id                    NaN         NaN   NaN    NaN    NaN    NaN     NaN  
order_item_id          1.197834    0.705124   1.0    1.0    1.0    1.0    21.0  
product_id                  NaN         NaN   NaN    NaN    NaN    NaN     NaN  
seller_id                   NaN         NaN   NaN    NaN    NaN    NaN     NaN  
shipping_limit_date         NaN         NaN   NaN    NaN    NaN    NaN     NaN  
price                120.653739  183.633928  0.85   39.9  74.99  134.9  6735.0  
freight_value          19.99032   15.806405   0.0  13.08  16.26  21.15  409.68


 [5] Missing Data


Columns  Missing Data
0             order_id             0
1        order_item_id             0
2           product_id             0
3            seller_id             0
4  shipping_limit_date             0
5                price             0
6        freight_value             0


 [6] First Rows


order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   
5  00048cc3ae777c65dbb7d2a0634bc1ea              1   
6  00054e8431b9d7675808bcb819fb4a32              1   
7  000576fe39319847cbb9d288c5617fa6              1   
8  0005a1a1728c9d785b8e2b08b904576c              1   
9  0005f50442cb953dcd1d21e1fb923495              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
5  ef92defde845ab8450f9d70c526ef70f  6426d21aca402a131fc0a5d0960a3c90   
6  8d4f2bb7e93e6710a28f34fa83ee7d28  7040e82f899a04d1b434b795a43b4617   
7  557d850972a7d6f792fd18ae1400d9b6  5996cddab893a4652a15592fb58ab8db   
8  310ae3c140ff94b03219ad0adc3c778f  a416b6a846a11724393025641d4edd5e   
9  4535b0e1091c278dfd193e5a1d63b39f  ba143b05f0110f0dc71ad71b4466ce92   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14  
5  2017-05-23 03:55:27   21.90          12.69  
6  2017-12-14 12:10:31   19.90          11.85  
7  2018-07-10 12:30:45  810.00          70.75  
8  2018-03-26 18:31:29  145.95          11.65  
9  2018-07-06 14:10:56   53.99          11.40


 [7] Last Rows


order_id  order_item_id  \
112640  fffb9224b6fc7c43ebb0904318b10b5f              1   
112641  fffb9224b6fc7c43ebb0904318b10b5f              2   
112642  fffb9224b6fc7c43ebb0904318b10b5f              3   
112643  fffb9224b6fc7c43ebb0904318b10b5f              4   
112644  fffbee3b5462987e66fb49b1c5411df2              1   
112645  fffc94f6ce00a00581880bf54a75a037              1   
112646  fffcd46ef2263f404302a634eb57f7eb              1   
112647  fffce4705a9662cd70adb13d4a31832d              1   
112648  fffe18544ffabc95dfada21779c9644f              1   
112649  fffe41c64501cc87c801fd61db3f6244              1   

                              product_id                         seller_id  \
112640  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   
112641  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   
112642  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   
112643  43423cdffde7fda63d0414ed38c11a73  b1fc4f64df5a0e8b6913ab38803c57a9   
112644  6f0169f259bb0ff432bfff7d829b9946  213b25e6f54661939f11710a6fddb871   
112645  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
112646  32e07fd915822b0765e448c4dd74c828  f3c38ab652836d21de61fb8314b69182   
112647  72a30483855e2eafc67aee5dc2560482  c3cfdc648177fdbbbb35635a37472c53   
112648  9c422a519119dcad7575db5af1ba540e  2b3e4a2a3ea8e01938cabda2a3e5cc79   
112649  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

        shipping_limit_date   price  freight_value  
112640  2017-11-03 02:55:58   55.00          34.19  
112641  2017-11-03 02:55:58   55.00          34.19  
112642  2017-11-03 02:55:58   55.00          34.19  
112643  2017-11-03 02:55:58   55.00          34.19  
112644  2018-06-28 09:58:03  119.85          20.03  
112645  2018-05-02 04:11:01  299.99          43.41  
112646  2018-07-20 04:31:48  350.00          36.53  
112647  2017-10-30 17:14:25   99.90          16.95  
112648  2017-08-21 00:04:32   55.99           8.72  
112649  2018-06-12 17:10:13   43.00          12.79


 [8] Summary


N_Rows  N_Columns    Memory
0  112650          7  37737882


Dataset: payment

 [1] Shape


N_Rows  N_columns
0  103886          5


 [2] Check Duplicated


np.int64(0)


 [3] Data Type


Columns Data Type
order_id                          order_id    object
payment_sequential      payment_sequential     int64
payment_type                  payment_type    object
payment_installments  payment_installments     int64
payment_value                payment_value   float64


 [4] Describe


count unique                               top  \
order_id                103886  99440  fa65dad1b0e818e3ccc5cb0e39231352   
payment_sequential    103886.0    NaN                               NaN   
payment_type            103886      5                       credit_card   
payment_installments  103886.0    NaN                               NaN   
payment_value         103886.0    NaN                               NaN   

                       freq       mean         std  min    25%    50%  \
order_id                 29        NaN         NaN  NaN    NaN    NaN   
payment_sequential      NaN   1.092679    0.706584  1.0    1.0    1.0   
payment_type          76795        NaN         NaN  NaN    NaN    NaN   
payment_installments    NaN   2.853349    2.687051  0.0    1.0    1.0   
payment_value           NaN  154.10038  217.494064  0.0  56.79  100.0   

                           75%       max  
order_id                   NaN       NaN  
payment_sequential         1.0      29.0  
payment_type               NaN       NaN  
payment_installments       4.0      24.0  
payment_value         171.8375  13664.08


 [5] Missing Data


Columns  Missing Data
0              order_id             0
1    payment_sequential             0
2          payment_type             0
3  payment_installments             0
4         payment_value             0


 [6] First Rows


order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   
5  298fcdf1f73eb413e4d26d01b25bc1cd                   1  credit_card   
6  771ee386b001f06208a7419e4fc1bbd7                   1  credit_card   
7  3d7239c394a212faae122962df514ac7                   1  credit_card   
8  1f78449c87a54faf9e96e88ba1491fa9                   1  credit_card   
9  0573b5e23cbd798006520e1d5b4c6714                   1       boleto   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45  
5                     2          96.12  
6                     1          81.16  
7                     3          51.84  
8                     6         341.09  
9                     1          51.95


 [7] Last Rows


order_id  payment_sequential payment_type  \
103876  0872d0faafe0cb56e6f2c594cca1522c                   1  credit_card   
103877  3531718b24dfd5022fb0d101e10e33b9                   1  credit_card   
103878  2e3d96522c8c378d4135c12a65c977aa                   1  credit_card   
103879  c45067032fd84f4cf408730ff5205568                   1  credit_card   
103880  7159096c5aa9be77f7f0c26c01ee9793                   1  credit_card   
103881  0406037ad97740d563a178ecc7a2075c                   1       boleto   
103882  7b905861d7c825891d6347454ea7863f                   1  credit_card   
103883  32609bbb3dd69b3c066a6860554a77bf                   1  credit_card   
103884  b8b61059626efa996a60be9bb9320e10                   1  credit_card   
103885  28bbae6599b09d39ca406b747b6632b1                   1       boleto   

        payment_installments  payment_value  
103876                     2         127.92  
103877                     1         194.11  
103878                     4          47.62  
103879                     2         198.94  
103880                     4         280.65  
103881                     1         363.31  
103882                     2          96.80  
103883                     1          47.77  
103884                     5         369.54  
103885                     1         191.58


 [8] Summary


N_Rows  N_Columns    Memory
0  103886          5  17017773


Dataset: oreder_reviews

 [1] Shape


N_Rows  N_columns
0   99224          7


 [2] Check Duplicated


np.int64(0)


 [3] Data Type


Columns Data Type
review_id                              review_id    object
order_id                                order_id    object
review_score                        review_score     int64
review_comment_title        review_comment_title    object
review_comment_message    review_comment_message    object
review_creation_date        review_creation_date    object
review_answer_timestamp  review_answer_timestamp    object


 [4] Describe


count unique                               top  \
review_id                  99224  98410  7b606b0d57b078384f0b58eac1d41d78   
order_id                   99224  98673  c88b1d1b157a9999ce368f218a407141   
review_score             99224.0    NaN                               NaN   
review_comment_title       11568   4527                         Recomendo   
review_comment_message     40977  36159                         Muito bom   
review_creation_date       99224    636               2017-12-19 00:00:00   
review_answer_timestamp    99224  98248               2017-06-15 23:21:05   

                        freq      mean       std  min  25%  50%  75%  max  
review_id                  3       NaN       NaN  NaN  NaN  NaN  NaN  NaN  
order_id                   3       NaN       NaN  NaN  NaN  NaN  NaN  NaN  
review_score             NaN  4.086421  1.347579  1.0  4.0  5.0  5.0  5.0  
review_comment_title     423       NaN       NaN  NaN  NaN  NaN  NaN  NaN  
review_comment_message   230       NaN       NaN  NaN  NaN  NaN  NaN  NaN  
review_creation_date     463       NaN       NaN  NaN  NaN  NaN  NaN  NaN  
review_answer_timestamp    4       NaN       NaN  NaN  NaN  NaN  NaN  NaN


 [5] Missing Data


Columns  Missing Data
0                review_id             0
1                 order_id             0
2             review_score             0
3     review_comment_title         87656
4   review_comment_message         58247
5     review_creation_date             0
6  review_answer_timestamp             0


 [6] First Rows


review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
5  15197aa66ff4d0650b5434f1b46cda19  b18dcdf73be66366873cd26c5724d1dc   
6  07f9bee5d1b850860defd761afa7ff16  e48aa0d2dcec3a2e87348811bcfdf22b   
7  7c6400515c67679fbee952a7525281ef  c31a859e34e3adac22f376954e19b39d   
8  a3f6f7f6f433de0aefbb97da197c554c  9c214ac970e84273583ab523dfafd09b   
9  8670d52e15e00043ae7de4c01cc2fe06  b9bf720beb4ab3728760088589c62129   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   
5             1                  NaN   
6             5                  NaN   
7             5                  NaN   
8             5                  NaN   
9             4            recomendo   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
5                                                NaN  2018-04-13 00:00:00   
6                                                NaN  2017-07-16 00:00:00   
7                                                NaN  2018-08-14 00:00:00   
8                                                NaN  2017-05-17 00:00:00   
9  aparelho eficiente. no site a marca do aparelh...  2018-05-22 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53  
5     2018-04-16 00:39:37  
6     2017-07-18 19:30:34  
7     2018-08-14 21:36:06  
8     2017-05-18 12:05:37  
9     2018-05-23 16:45:47


 [7] Last Rows


review_id                          order_id  \
99214  3c569245c7610d9a51eff50032194ea3  bf25ff262495b0b876bc4f594f736ecf   
99215  a709d176f59bc3af77f4149c96bae357  d5cb12269711bd1eaf7eed8fd32a7c95   
99216  115d72425fc59eec9a2e9d7594c6d93d  acd45245723df7cb52772a34416b41b1   
99217  c6b270c61f67c9f7cb07d84ea8aeaf8b  48f7ee67313eda32bfcf5b9c1dd9522d   
99218  af2dc0519de6e0720ef0c74292fb4114  d699c734a0b1c8111f2272a3f36d398c   
99219  574ed12dd733e5fa530cfd4bbf39d7c9  2a8c23fee101d4d5662fa670396eb8da   
99220  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99221  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99222  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99223  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
99214             5                  NaN   
99215             3                  NaN   
99216             3                  NaN   
99217             5                  NaN   
99218             5                  NaN   
99219             5                  NaN   
99220             5                  NaN   
99221             5                  NaN   
99222             4                  NaN   
99223             1                  NaN   

                                  review_comment_message review_creation_date  \
99214                                                NaN  2018-05-23 00:00:00   
99215  O produto não foi enviado com NF, não existe v...  2018-05-19 00:00:00   
99216                                                NaN  2018-05-08 00:00:00   
99217                                                NaN  2017-12-13 00:00:00   
99218                                                NaN  2018-04-27 00:00:00   
99219                                                NaN  2018-07-07 00:00:00   
99220                                                NaN  2017-12-09 00:00:00   
99221  Excelente mochila, entrega super rápida. Super...  2018-03-22 00:00:00   
99222                                                NaN  2018-07-01 00:00:00   
99223  meu produto chegou e ja tenho que devolver, po...  2017-07-03 00:00:00   

      review_answer_timestamp  
99214     2018-05-26 21:24:06  
99215     2018-05-20 21:51:06  
99216     2018-05-10 14:37:03  
99217     2017-12-14 11:09:36  
99218     2018-04-30 01:18:57  
99219     2018-07-14 17:18:30  
99220     2017-12-11 20:06:42  
99221     2018-03-23 09:10:43  
99222     2018-07-02 12:59:13  
99223     2017-07-03 21:01:49


 [8] Summary


N_Rows  N_Columns    Memory
0   99224          7  41025302


Dataset: orders

 [1] Shape


N_Rows  N_columns
0   99441          8


 [2] Check Duplicated


np.int64(0)


 [3] Data Type


Columns Data Type
order_id                                            order_id    object
customer_id                                      customer_id    object
order_status                                    order_status    object
order_purchase_timestamp            order_purchase_timestamp    object
order_approved_at                          order_approved_at    object
order_delivered_carrier_date    order_delivered_carrier_date    object
order_delivered_customer_date  order_delivered_customer_date    object
order_estimated_delivery_date  order_estimated_delivery_date    object


 [4] Describe


count unique                               top  \
order_id                       99441  99441  e481f51cbdc54678b7cc49136f2d6af7   
customer_id                    99441  99441  9ef432eb6251297304e76186b10a928d   
order_status                   99441      8                         delivered   
order_purchase_timestamp       99441  98875               2018-04-11 10:48:14   
order_approved_at              99281  90733               2018-02-27 04:31:10   
order_delivered_carrier_date   97658  81018               2018-05-09 15:48:00   
order_delivered_customer_date  96476  95664               2018-05-08 23:38:46   
order_estimated_delivery_date  99441    459               2017-12-20 00:00:00   

                                freq  
order_id                           1  
customer_id                        1  
order_status                   96478  
order_purchase_timestamp           3  
order_approved_at                  9  
order_delivered_carrier_date      47  
order_delivered_customer_date      3  
order_estimated_delivery_date    522


 [5] Missing Data


Columns  Missing Data
0                       order_id             0
1                    customer_id             0
2                   order_status             0
3       order_purchase_timestamp             0
4              order_approved_at           160
5   order_delivered_carrier_date          1783
6  order_delivered_customer_date          2965
7  order_estimated_delivery_date             0


 [6] First Rows


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
5  a4591c265e18cb1dcee52889e2d8acc3  503740e9ca751ccdda7ba28e9ab8f608   
6  136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
7  6514b8ad8028c9f2cc2374ded245783f  9bdf08b4b3b52b5526ff42d37d47f222   
8  76c6e866289321a7c93b82b54852dc33  f54a9f0e6b351c431402b8461ea51999   
9  e69bfb5eb88e0ed6a785585b27e16dbf  31ad1d1b63eb9962463f764d4e6e0c9d   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
5    delivered      2017-07-09 21:57:05  2017-07-09 22:10:13   
6     invoiced      2017-04-11 12:22:08  2017-04-13 13:25:17   
7    delivered      2017-05-16 13:10:30  2017-05-16 13:22:11   
8    delivered      2017-01-23 18:29:09  2017-01-25 02:50:47   
9    delivered      2017-07-29 11:55:02  2017-07-29 12:05:32   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   
5          2017-07-11 14:58:04           2017-07-26 10:57:55   
6                          NaN                           NaN   
7          2017-05-22 10:07:46           2017-05-26 12:55:51   
8          2017-01-26 14:16:31           2017-02-02 14:08:10   
9          2017-08-10 19:45:24           2017-08-16 17:14:30   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00  
5           2017-08-01 00:00:00  
6           2017-05-09 00:00:00  
7           2017-06-07 00:00:00  
8           2017-03-06 00:00:00  
9           2017-08-23 00:00:00


 [7] Last Rows


order_id                       customer_id  \
99431  b0f4af5c1b06e24fef510703bfe9f0a6  8e1ec396e317ff4c82a03ce16a0c3eb3   
99432  cfa78b997e329a5295b4ee6972c02979  a2f7428f0cafbc8e59f20e1444b67315   
99433  9115830be804184b91f5c00f6f49f92d  da2124f134f5dfbce9d06f29bdb6c308   
99434  aa04ef5214580b06b10e2a378300db44  f01a6bfcc730456317e4081fe0c9940e   
99435  880675dff2150932f1601e1c07eadeeb  47cd45a6ac7b9fb16537df2ccffeb5ac   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp    order_approved_at  \
99431    delivered      2017-10-27 15:21:00  2017-10-27 15:32:49   
99432    delivered      2017-12-20 09:52:41  2017-12-20 10:09:52   
99433    delivered      2017-10-04 19:57:37  2017-10-04 20:07:14   
99434    delivered      2017-01-27 00:30:03  2017-01-27 01:05:25   
99435    delivered      2017-02-23 09:05:12  2017-02-23 09:15:11   
99436    delivered      2017-03-09 09:54:05  2017-03-09 09:54:05   
99437    delivered      2018-02-06 12:58:58  2018-02-06 13:10:37   
99438    delivered      2017-08-27 14:46:43  2017-08-27 15:04:16   
99439    delivered      2018-01-08 21:28:27  2018-01-08 21:36:21   
99440    delivered      2018-03-08 20:57:30  2018-03-09 11:20:28   

      order_delivered_carrier_date order_delivered_customer_date  \
99431          2017-10-30 15:44:34           2017-11-10 17:57:22   
99432          2017-12-20 20:25:25           2018-01-26 15:45:14   
99433          2017-10-05 16:52:52           2017-10-20 20:25:45   
99434          2017-01-30 11:40:16           2017-02-07 13:15:25   
99435          2017-03-01 10:22:52           2017-03-06 11:08:08   
99436          2017-03-10 11:18:03           2017-03-17 15:08:01   
99437          2018-02-07 23:22:42           2018-02-28 17:37:56   
99438          2017-08-28 20:52:26           2017-09-21 11:24:17   
99439          2018-01-12 15:35:03           2018-01-25 23:32:54   
99440          2018-03-09 22:11:59           2018-03-16 13:08:30   

      order_estimated_delivery_date  
99431           2017-11-22 00:00:00  
99432           2018-01-18 00:00:00  
99433           2017-11-07 00:00:00  
99434           2017-03-17 00:00:00  
99435           2017-03-22 00:00:00  
99436           2017-03-28 00:00:00  
99437           2018-03-02 00:00:00  
99438           2017-09-27 00:00:00  
99439           2018-02-15 00:00:00  
99440           2018-04-03 00:00:00


 [8] Summary


N_Rows  N_Columns    Memory
0   99441          8  55508758


Dataset: products

 [1] Shape


N_Rows  N_columns
0   32951          9


 [2] Check Duplicated


np.int64(0)


 [3] Data Type


Columns Data Type
product_id                                  product_id    object
product_category_name            product_category_name    object
product_name_lenght                product_name_lenght   float64
product_description_lenght  product_description_lenght   float64
product_photos_qty                  product_photos_qty   float64
product_weight_g                      product_weight_g   float64
product_length_cm                    product_length_cm   float64
product_height_cm                    product_height_cm   float64
product_width_cm                      product_width_cm   float64


 [4] Describe


count unique                               top  \
product_id                    32951  32951  1e9e8ef04dbcff4541ed26657ea517e5   
product_category_name         32341     73                   cama_mesa_banho   
product_name_lenght         32341.0    NaN                               NaN   
product_description_lenght  32341.0    NaN                               NaN   
product_photos_qty          32341.0    NaN                               NaN   
product_weight_g            32949.0    NaN                               NaN   
product_length_cm           32949.0    NaN                               NaN   
product_height_cm           32949.0    NaN                               NaN   
product_width_cm            32949.0    NaN                               NaN   

                            freq         mean          std  min    25%    50%  \
product_id                     1          NaN          NaN  NaN    NaN    NaN   
product_category_name       3029          NaN          NaN  NaN    NaN    NaN   
product_name_lenght          NaN    48.476949    10.245741  5.0   42.0   51.0   
product_description_lenght   NaN   771.495285   635.115225  4.0  339.0  595.0   
product_photos_qty           NaN     2.188986     1.736766  1.0    1.0    1.0   
product_weight_g             NaN  2276.472488  4282.038731  0.0  300.0  700.0   
product_length_cm            NaN    30.815078    16.914458  7.0   18.0   25.0   
product_height_cm            NaN    16.937661    13.637554  2.0    8.0   13.0   
product_width_cm             NaN    23.196728    12.079047  6.0   15.0   20.0   

                               75%      max  
product_id                     NaN      NaN  
product_category_name          NaN      NaN  
product_name_lenght           57.0     76.0  
product_description_lenght   972.0   3992.0  
product_photos_qty             3.0     20.0  
product_weight_g            1900.0  40425.0  
product_length_cm             38.0    105.0  
product_height_cm             21.0    105.0  
product_width_cm              30.0    118.0


 [5] Missing Data


Columns  Missing Data
0                  product_id             0
1       product_category_name           610
2         product_name_lenght           610
3  product_description_lenght           610
4          product_photos_qty           610
5            product_weight_g             2
6           product_length_cm             2
7           product_height_cm             2
8            product_width_cm             2


 [6] First Rows


product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   
5  41d3672d4792049fa1779bb35283ed13  instrumentos_musicais   
6  732bd381ad09e530fe0a5f457d81becb             cool_stuff   
7  2548af3e6e77a690cf3eb6368e9ab61e       moveis_decoracao   
8  37cc742be07708b53a98702e77a21a02       eletrodomesticos   
9  8c92109888e8cdf9d66dc7e463025574             brinquedos   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   
5                 60.0                       745.0                 1.0   
6                 56.0                      1272.0                 4.0   
7                 56.0                       184.0                 2.0   
8                 57.0                       163.0                 1.0   
9                 36.0                      1156.0                 1.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0  
5             200.0               38.0                5.0              11.0  
6           18350.0               70.0               24.0              44.0  
7             900.0               40.0                8.0              40.0  
8             400.0               27.0               13.0              17.0  
9             600.0               17.0               10.0              12.0


 [7] Last Rows


product_id              product_category_name  \
32941  6ec96c91757fad0aecafc0ee7f262dcc                              bebes   
32942  16280ca280a86fee2ba3c928ed04439f                   moveis_decoracao   
32943  3becff10d1deb92b02f2a1ee62a04524             informatica_acessorios   
32944  1a14237ecc2fe3772b55c8d4e11ccb35                   moveis_decoracao   
32945  c4e71b64511b959455e2107fe7859020              utilidades_domesticas   
32946  a0b7d5a992ccda646f2d34e418fff5a0                   moveis_decoracao   
32947  bf4538d88321d0fd4412a93c974510e6  construcao_ferramentas_iluminacao   
32948  9a7c6041fa9592d9d9ef6cfe62a71f8c                    cama_mesa_banho   
32949  83808703fc0706a22e264b9d75f04a2e             informatica_acessorios   
32950  106392145fca363410d287a815be6de4                    cama_mesa_banho   

       product_name_lenght  product_description_lenght  product_photos_qty  \
32941                 62.0                      1417.0                 1.0   
32942                 64.0                       236.0                11.0   
32943                 54.0                      1520.0                 2.0   
32944                 58.0                      1405.0                 3.0   
32945                 59.0                      1371.0                 2.0   
32946                 45.0                        67.0                 2.0   
32947                 41.0                       971.0                 1.0   
32948                 50.0                       799.0                 1.0   
32949                 60.0                       156.0                 2.0   
32950                 58.0                       309.0                 1.0   

       product_weight_g  product_length_cm  product_height_cm  \
32941            9550.0               36.0               35.0   
32942            2200.0               31.0               11.0   
32943            6150.0               30.0               30.0   
32944             150.0               35.0                2.0   
32945             200.0               18.0               15.0   
32946           12300.0               40.0               40.0   
32947            1700.0               16.0               19.0   
32948            1400.0               27.0                7.0   
32949             700.0               31.0               13.0   
32950            2083.0               12.0                2.0   

       product_width_cm  
32941              35.0  
32942              26.0  
32943              20.0  
32944              25.0  
32945              15.0  
32946              40.0  
32947              16.0  
32948              27.0  
32949              20.0  
32950               7.0


 [8] Summary


N_Rows  N_Columns   Memory
0   32951          9  6602426


Dataset: product_category

 [1] Shape


N_Rows  N_columns
0      71          2


 [2] Check Duplicated


np.int64(0)


 [3] Data Type


Columns Data Type
product_category_name                  product_category_name    object
product_category_name_english  product_category_name_english    object


 [4] Describe


count unique            top freq
product_category_name            71     71   beleza_saude    1
product_category_name_english    71     71  health_beauty    1


 [5] Missing Data


Columns  Missing Data
0          product_category_name             0
1  product_category_name_english             0


 [6] First Rows


product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor
5           esporte_lazer                sports_leisure
6              perfumaria                     perfumery
7   utilidades_domesticas                    housewares
8               telefonia                     telephony
9      relogios_presentes                 watches_gifts


 [7] Last Rows


product_category_name          product_category_name_english
61                         musica                                  music
62                casa_conforto_2                         home_comfort_2
63    portateis_casa_forno_e_cafe  small_appliances_home_oven_and_coffee
64              cds_dvds_musicais                      cds_dvds_musicals
65                   dvds_blu_ray                           dvds_blu_ray
66                         flores                                flowers
67             artes_e_artesanato                  arts_and_craftmanship
68                fraldas_higiene                    diapers_and_hygiene
69  fashion_roupa_infanto_juvenil              fashion_childrens_clothes
70             seguros_e_servicos                  security_and_services


 [8] Summary


N_Rows  N_Columns  Memory
0      71          2    9436


Dataset: sellers

 [1] Shape


N_Rows  N_columns
0    3095          4


 [2] Check Duplicated


np.int64(0)


 [3] Data Type


Columns Data Type
seller_id                            seller_id    object
seller_zip_code_prefix  seller_zip_code_prefix     int64
seller_city                        seller_city    object
seller_state                      seller_state    object


 [4] Describe


count unique                               top  freq  \
seller_id                 3095   3095  3442f8959a84dea7ee197c632cb2df15     1   
seller_zip_code_prefix  3095.0    NaN                               NaN   NaN   
seller_city               3095    611                         sao paulo   694   
seller_state              3095     23                                SP  1849   

                                mean          std     min     25%      50%  \
seller_id                        NaN          NaN     NaN     NaN      NaN   
seller_zip_code_prefix  32291.059451  32713.45383  1001.0  7093.5  14940.0   
seller_city                      NaN          NaN     NaN     NaN      NaN   
seller_state                     NaN          NaN     NaN     NaN      NaN   

                            75%      max  
seller_id                   NaN      NaN  
seller_zip_code_prefix  64552.5  99730.0  
seller_city                 NaN      NaN  
seller_state                NaN      NaN


 [5] Missing Data


Columns  Missing Data
0               seller_id             0
1  seller_zip_code_prefix             0
2             seller_city             0
3            seller_state             0


 [6] First Rows


seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   
5  c240c4061717ac1806ae6ee72be3533b                   20920   
6  e49c26c3edfa46d227d5121a6b6e4d37                   55325   
7  1b938a7ec6ac5061a66a3766e0e75f90                   16304   
8  768a86e36ad6aae3d03ee3c6433d61df                    1529   
9  ccc4bbb5f32a6ab2b7066a4130f114e3                   80310   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP  
5     rio de janeiro           RJ  
6             brejao           PE  
7          penapolis           SP  
8          sao paulo           SP  
9           curitiba           PR


 [7] Last Rows


seller_id  seller_zip_code_prefix seller_city  \
3085  9d489893ff6af05029671b785f54b998                   13613        leme   
3086  725af026f869236a8dd0ceb98564d156                   35530     claudio   
3087  f00f5b35d0abcacbdd863672f4bb2c1a                    1238   sao paulo   
3088  7bac63f6603d382cc8d0832eb6c100a8                    5713   sao paulo   
3089  f1fdf2d13186575751aa25876536d85c                    5314   sao paulo   
3090  98dddbc4601dd4443ca174359b237166                   87111     sarandi   
3091  f8201cab383e484733266d1906e2fdfa                   88137     palhoca   
3092  74871d19219c7d518d0090283e03c137                    4650   sao paulo   
3093  e603cf3fec55f8697c9059638d6c8eb5                   96080     pelotas   
3094  9e25199f6ef7e7c347120ff175652c3b                   12051     taubate   

     seller_state  
3085           SP  
3086           MG  
3087           SP  
3088           SP  
3089           SP  
3090           PR  
3091           SC  
3092           SP  
3093           RS  
3094           SP


 [8] Summary


N_Rows  N_Columns  Memory
0    3095          4  616671

# Key Data Quality Issues Summary – Olist Dataset

After initial exploration, here are the **most important observations** that require attention during the data cleaning phase:

## 1. Date / Timestamp Columns (Critical)
- **All date-related columns are stored as object (string)** instead of datetime  
  → Needs urgent conversion using `pd.to_datetime()`
- Affected columns (partial list):
  - orders: `order_purchase_timestamp`, `order_approved_at`, `order_delivered_carrier_date`, `order_delivered_customer_date`, `order_estimated_delivery_date`
  - order_items: `shipping_limit_date`
  - order_reviews: `review_creation_date`, `review_answer_timestamp`
- **Impact**: Cannot perform time-based analysis, calculations, or sorting correctly until converted

## 2. Missing Values (Missing Data)
| Table                     | Column(s) with missing values                  | Approx. missing count / %         | Priority    | Recommendation                              |
|---------------------------|------------------------------------------------|-----------------------------------|-------------|---------------------------------------------|
| orders                    | order_approved_at                              | 160                               | Medium      | Impute with purchase timestamp or drop rows |
| orders                    | order_delivered_carrier_date                   | ~1,783                            | High        | Important for delivery analysis             |
| orders                    | order_delivered_customer_date                  | ~2,965 (~3%)                      | High        | Critical for delivery time calculation      |
| products                  | product_category_name, name_lenght, description_lenght, photos_qty | 610 (~1.85%)     | Medium      | Impute category as 'unknown'                |
| products                  | weight_g, dimensions                           | 2                                 | Low         | Impute with median                          |
| order_reviews             | review_comment_title                           | ~87,656 (~88%)                    | High        | Almost always missing → drop or flag        |
| order_reviews             | review_comment_message                         | ~58,247 (~59%)                    | High        | High missing rate → consider text analysis only on available |

## 3. Duplicates / Uniqueness Issues
- **No critical duplicates** in primary keys across most tables (customer_id, order_id, product_id, seller_id, review_id are unique)
- **Major issue in geolocation table**:
  - `geolocation_zip_code_prefix` has **severe duplication** (~261,831 duplicates)
  - Cannot be used as a clean primary key or direct join key
  - **Solution already applied**: Created aggregated table `zipcode_locations` (unique zip prefix + averages)

## 4. Other Important Observations
- `customer_unique_id` is the **real customer identifier** (not `customer_id`)
  → Use it when analyzing customer behavior across multiple orders
- `product_category_name` in products has ~610 missing → must be handled before joining with translation table
- All text-based IDs (`object` dtype) are clean (no missing values)
- No negative values reported in price, freight_value, weight, etc. (still good to double-check)

## Summary of Priority Cleaning Actions
1. **Convert all timestamp columns** to proper datetime type
2. **Handle missing values** in delivery dates (orders) and review comments
3. **Use aggregated zipcode_locations** instead of raw geolocation table
4. **Impute or flag** missing product category and dimensions
5. **Consider dropping** or special handling for columns with very high missing rate (e.g., review_comment_title)

These are the **main pain points** that will affect analysis quality if not addressed.

## **Data Cleaning**

In [ ]:
# Handling Geolocation Duplicates: Removing exact duplicates to optimize the dataset
geolocation.drop_duplicates(inplace= True)

In [ ]:
# Data Type Correction: Converting string objects to datetime for time-series analysis
items.shipping_limit_date =             pd.to_datetime(items.shipping_limit_date)
o_reviews.review_creation_date =        pd.to_datetime(o_reviews.review_creation_date)
o_reviews.review_answer_timestamp =     pd.to_datetime(o_reviews.review_answer_timestamp)

# Bulk conversion for 'orders' table timestamps
date_cols = ['order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 
             'order_delivered_customer_date', 'order_estimated_delivery_date']
for col in date_cols:
    orders[col] = pd.to_datetime(orders[col], errors= 'coerce')

In [ ]:
# Handling Missing Values: Imputing missing comments and category info
o_reviews.review_comment_title.fillna('no_comment', inplace= True)
o_reviews.review_comment_message.fillna('no_comment', inplace= True)

# Imputing product categories with the most frequent value (Mode)
if products.product_category_name.isnull().any():
    products.product_category_name.fillna(products.product_category_name.mode()[0], inplace=True)

# Imputing numerical product specs with Median (robust to outliers)
num_cols = ['product_name_lenght', 'product_description_lenght', 'product_photos_qty']
for col in num_cols:
    products[col].fillna(products[col].median(), inplace=True)

C:\Users\Ali\AppData\Local\Temp\ipykernel_8904\4278697606.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  o_reviews.review_comment_title.fillna('no_comment', inplace= True)
C:\Users\Ali\AppData\Local\Temp\ipykernel_8904\4278697606.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a co

In [ ]:
# Exporting cleaned datasets to CSV for the next stage (SQL/Power BI)
customers.to_csv('customers.csv', index= False)
geolocation.to_csv('geolocation.csv', index= False)
items.to_csv('order_items.csv', index= False)
payment.to_csv('order_payment.csv', index= False)
o_reviews.to_csv('order_review.csv', index= False)
orders.to_csv('orders.csv', index= False)
products.to_csv('products.csv', index= False)
product_category.to_csv('product_category.csv', index= False)
sellers.to_csv('sellers.csv', index= False)